core/transaction/payload/Subscribe.go

```diff
@@ -0,0 +1,127 @@
+package payload
+
+import (
+	"bytes"
+	"encoding/json"
+	"io"
+	"strings"
+
+	"github.com/nknorg/nkn/common"
+	"github.com/nknorg/nkn/common/serialization"
+	. "github.com/nknorg/nkn/errors"
+)
+
+type Subscribe struct {
+	Subscriber []byte
+	Identifier string
+	Topic      string
+	Duration   uint32
+}
+
+func (s *Subscribe) Data(version byte) []byte {
+	//TODO: implement Subscribe.Data()
+	return []byte{0}
+
+}
+
+func (s *Subscribe) Serialize(w io.Writer, version byte) error {
+	serialization.WriteVarBytes(w, s.Subscriber)
+	serialization.WriteVarString(w, s.Identifier)
+	serialization.WriteVarString(w, s.Topic)
+	serialization.WriteUint32(w, s.Duration)
+	return nil
+}
+
+func (s *Subscribe) Deserialize(r io.Reader, version byte) error {
+	var err error
+	s.Subscriber, err = serialization.ReadVarBytes(r)
+	if err != nil {
+		return NewDetailErr(err, ErrNoCode, "[Subscribe], Subscriber Deserialize failed.")
+	}
+	s.Identifier, err = serialization.ReadVarString(r)
+	if err != nil {
+		return NewDetailErr(err, ErrNoCode, "[Subscribe], Identifier Deserialize failed.")
+	}
+	s.Topic, err = serialization.ReadVarString(r)
+	if err != nil {
+		return NewDetailErr(err, ErrNoCode, "[Subscribe], Topic Deserialize failed.")
+	}
+	s.Duration, err = serialization.ReadUint32(r)
+	if err != nil {
+		return NewDetailErr(err, ErrNoCode, "[Subscribe], Duration Deserialize failed.")
+	}
+	return nil
+}
+
+func (s *Subscribe) Equal(s2 *Subscribe) bool {
+	if !bytes.Equal(s.Subscriber, s2.Subscriber) {
+		return false
+	}
+
+	if s.Identifier != s2.Identifier {
+		return false
+	}
+
+	if s.Topic != s2.Topic {
+		return false
+	}
+
+	if s.Duration != s2.Duration {
+		return false
+	}
+
+	return true
+}
+
+func (s *Subscribe) SubscriberString() string {
+	return SubscriberString(s.Subscriber, s.Identifier)
+}
+
+func SubscriberString(subscriber []byte, identifier string) string {
+	var result strings.Builder
+	if identifier != "" {
+		result.WriteString(identifier)
+		result.WriteString(".")
+	}
+	result.WriteString(common.BytesToHexString(subscriber))
+
+	return result.String()
+}
+
+func (s *Subscribe) MarshalJson() ([]byte, error) {
+	si := &SubscribeInfo{
+		Subscriber: s.SubscriberString(),
+		Topic:      s.Topic,
+		Duration:   s.Duration,
+	}
+
+	data, err := json.Marshal(si)
+	if err != nil {
+		return nil, err
+	}
+	return data, nil
+}
+
+func (s *Subscribe) UnmarshalJson(data []byte) error {
+	si := new(SubscribeInfo)
+	var err error
+	if err = json.Unmarshal(data, &si); err != nil {
+		return err
+	}
+
+	substrings := strings.Split(si.Subscriber, ".")
```

Comment:
What about move the client address encode/decode part into somewhere (e.g. util/address/address.go) so that all address scheme code are in the same module and other (future) scheme-related code can also use it?

Reply:
done